src/shared/utils/fluxData.ts

```diff
@@ -0,0 +1,135 @@
+import {FLUX_RESPONSE_BYTES_LIMIT} from 'src/shared/constants'
+import {getFlagValue} from 'src/shared/utils/featureFlag'
+import {PROJECT_NAME} from 'src/flows/constants'
+import {FluxResult, FluxResultComplete} from 'src/shared/apis/query'
+
+/*
+  Given an arbitrary text chunk of a Flux CSV, trim partial lines off of the end
+  of the text.
+
+  For example, given the following partial Flux response,
+
+            r,baz,3
+      foo,bar,baz,2
+      foo,bar,b
+
+  we want to trim the last incomplete line, so that the result is
+
+            r,baz,3
+      foo,bar,baz,2
+
+*/
+export const trimPartialLines = (partialResp: string): string => {
+  let i = partialResp.length - 1
+
+  while (partialResp[i] !== '\n') {
+    if (i <= 0) {
+      return partialResp
+    }
+
+    i -= 1
+  }
+
+  return partialResp.slice(0, i + 1)
+}
+
+export const extractTableId = (str: string): number | null => {
+  let prevIdx = 0
+  let currIdx = 0
+  let col = 0
+  let maxTableIdx = null
+
+  const nextLine = () => {
+    col = 0
+    prevIdx = currIdx
+    const nextIdx = str.substring(currIdx + 1).search(/\r\n/)
+    currIdx = nextIdx == -1 ? nextIdx : currIdx + 1 + nextIdx + 2
+  }
+  const nextCol = () => {
+    prevIdx = currIdx
+    const nextIdx = str.substring(currIdx + 1).search(/\,/)
+    currIdx = nextIdx == -1 ? nextIdx : currIdx + 1 + nextIdx
+  }
+
+  nextLine()
+  while (currIdx !== -1 && str[currIdx] !== undefined) {
+    switch (str[currIdx]) {
+      case '#':
+        nextLine()
+        break
+      case ',':
+        if (col == 2) {
+          try {
+            const tableIdx = parseInt(str.substring(prevIdx + 1, currIdx))
+            maxTableIdx = Math.max(maxTableIdx || 0, tableIdx)
+          } catch {}
+          nextLine()
+          break
+        }
+        col++
+        nextCol()
+        break
+      case '\r':
+      case '\n':
+      default:
+        nextLine()
+        break
+    }
+  }
+  return maxTableIdx
+}
+
+export async function* fluxDataReader(
+  responseBody: ReadableStream<Uint8Array>
+): AsyncGenerator<FluxResult, FluxResultComplete> {
```

Comment:
The return signature is very intentional. Such that the consumers of this function do not need to care if the data was truncated (with 2 yields) or not (with only 1 yield). That is why the yield signatures are identical except for the extended `FluxResultComplete` type includes the metadata.